### PR TITLE
Add suggested changes from API guild discussion (POST usage)

### PR DESCRIPTION
**Pull request by SmilyOrg**
I took some liberties on the formatting/wording, but mostly it's as discussed in the [internal doc](https://docs.google.com/document/d/1NahSuMC0LRwFTTrln2m0iDLq0nHzGmF59Gv7xBhKPfQ/edit#bookmark=id.e983u51x2ler).

### CODE REVIEW

File: chapters/http-requests.adoc [hunk 1]
```diff
@@ -122,38 +122,65 @@ details and options.
...
+See below for common {POST} use-cases.
+
+[[post-create-single]]
+==== Single resource creation
+
+* Successful {POST} requests on a collection endpoint {SHOULD} return {201} and
+  the new resource object (including the resource identifier) in the response
+  payload.
+** The client {SHOULD} be able to use this resource identifier to {GET} the
+   resource if supported by the API, see <<143>>.
```
Comment (tkrop) [inline chapters/http-requests.adoc:146]:
```suggestion
** The client {SHOULD} be able to use this resource identifier to {GET} the
   resource if supported by the API (see <<143>>).
```

File: chapters/http-requests.adoc [hunk 2]
```diff
@@ -122,38 +122,65 @@ details and options.
...
+* If the response does not contain the created resource (no payload, or a
+  payload with just a "Success" message), the status code should still be {201},
+  and the URL to {GET} the resource {MUST} be provided in the {Location}
+  response header. It is a good practice to always provide the {Location} header
+  with the URL to the created resource.
```
Comment (tkrop) [inline chapters/http-requests.adoc:151]: I would not like to advise to send just "success" message as response. This is not a resourcish approach as should be targeted by restful APIs.

* A `POST` request always creates or updates a resource or optionally a resource monitor.

The only degree of freedom in the API is to decide, whether the "resource" (or a "resource monitor") or nothing is returned, and whether a `Location` header or no `Location` header is returned. Suggesting you could create just a "success" message is leading in the wrong direction in my opinion.

```suggestion
* If the response does not contain the created resource or a resource monitor the
  status code should still be {201}, and the URL to {GET} the resource {MUST} be
  provided in the {Location} response header. It is a good practice to always provide
  the {Location} header with the URL to the created resource (see <<180>>).
```

**Comment by tkrop** [comment]
I'm a bit hesitant to approve this, because the pull request majorly splits up the dense description from before into 4 subsections without providing additional clarification that wasn't clear from the before context.

The wording improvements are fine and could be added to the before structure too.

File: chapters/http-requests.adoc (hunk 2 above)
Comment (SmilyOrg) [reply chapters/http-requests.adoc:151]: Yeah, makes sense.

File: chapters/http-requests.adoc [hunk 3]
```diff
@@ -122,38 +122,65 @@ details and options.
...
-* Successful {POST} requests return {200} or {204} (if the resource was updated
-  -- with or without returning the resource), {201} (if the resource was newly
-  created), and {202} (if the request was accepted for asynchronous
-  processing).
+are equally possible.
 
-*Note:* By using {POST} to create resources the resource ID must not be passed
-as request input date by the client, but created and maintained by the service
-and returned with the response payload.
+The semantic for collection endpoints is best described as _"please add the
```
[the remark this answers is not in the document]
Comment (SmilyOrg) [reply chapters/http-requests.adoc]: Removed

**Comment by SmilyOrg** [comment]
> I'm a bit hesitant to approve this, because the pull request majorly splits up the dense description from before into 4 subsections without providing additional clarification that wasn't clear from the before context. In my opinion the deeper nesting is not preferable here.
> 
> The wording improvements are fine and could be added to the before structure too.

You're right, looking at it again it seems a bit too wordy. However, the previous dense description seemed a bit too overloaded, so there was a lot to interpret. Maybe there's a good structure somewhere in the middle? I'll have to take another look :)

Either way, I applied the suggestions, thanks!

**Comment by tfrauenstein** [comment]
👍

File: chapters/http-requests.adoc [hunk 4]
```diff
@@ -117,73 +122,106 @@ details and options.
...
+** The client {SHOULD} be able to use this resource identifier to {GET} the
+   resource if supported by the API (see <<143>>).
+* If the response does not contain the created resource or a resource monitor the
+  status code should still be {201}, and the URL to {GET} the resource {MUST} be
+  provided in the {Location} response header. It is a good practice to always provide
+  the {Location} header with the URL to the created resource (see <<180>>).
+* The resource ID {MUST-NOT} be passed as request body data by the client,
+  but created and maintained by the service.
+* Posting the same resource twice is *not* required to be <<idempotent>>
+and may result in multiple resources, see <<post-idempotent>>.
```
Comment (tkrop) [inline chapters/http-requests.adoc:153]: This is already stated in the hint in the main section, so I think we should skip it here.

File: chapters/http-requests.adoc [hunk 5]
```diff
@@ -117,73 +122,106 @@ details and options.
...
+* If the resource was created, the response {SHOULD} have a {201} status code
+  and follow <<post-create-single>> semantics.
+* If the resource was updated, the response status code should be {200} or
+  {204} (with or without returning the resource).
+
+[[post-async]]
+==== Asynchronous processing
+
+If the request is accepted for asynchronous processing, the response status code
+should be {202}, see <<253>>.
```
Comment (tkrop) [inline chapters/http-requests.adoc:181]: This sound somehow over-separated. It is only one sentence and effectively can be applied to all of the before use cases. I vote for adding the `202` response code as options to the before three use case that already distinguish different `2xx` response codes in each case. `202` would just be another one in the list.

File: chapters/http-status-codes-and-errors.adoc [hunk 6]
```diff
@@ -169,60 +167,54 @@ cases also for {DELETE}.
...
-{MUST} change the method to {GET}.
-
+The response to the request can be found under another URI using a {GET}
+method. A disambiguated version of {302} for the case where the client {MUST}
+change the method to {GET}. See also <<251>>.
 
 [[status-code-304]]
 ==== 304 Not Modified {rfc-status-304} {document} {GET} {HEAD}
 [.indent]
 Indicates that a conditional {GET} or {HEAD} request would have resulted in
```
Comment (tkrop) [inline chapters/http-status-codes-and-errors.adoc:196]:
```suggestion
Indicates that a conditional {GET} or {HEAD} request would have resulted in a
```

File: chapters/http-status-codes-and-errors.adoc [hunk 7]
```diff
@@ -169,60 +167,54 @@ cases also for {DELETE}.
...
-
 [.do-not-use]
 [[status-code-301]]
 ==== 301 Moved Permanently {rfc-status-301} {do-not-use} {ALL}
 [.indent]
-This and all future requests should be directed to the
-given URI.
-
+This and all future requests should be directed to the given URI. See also
+<<251>>.
```
Comment (tkrop) [inline chapters/http-status-codes-and-errors.adoc:175]: The link to rule `250` is surely wrong, but I still think, that adding the reference to `251` on the top is shorter and more effective than adding the reference to each effected `3xx` response code. The exception of `304` in my opinion is marked sufficiently be the rule.

File: chapters/http-status-codes-and-errors.adoc [hunk 8]
```diff
@@ -366,53 +345,102 @@ The server understands the content type, but is unable to process the content.
...
+==== 501 Not Implemented {rfc-status-501} {document} {ALL}
+[.indent]
+Server cannot fulfill the request, since the endpoint is not implemented yet.
+Usually this implies future availability, but retrying now is not recommended.
+May be documented on endpoints that are planned to be implemented in the
+future to indicate that they are still not available.
+
+[[status-code-502]]
+==== 502 Bad Gateway {rfc-status-502} {ALL}
+[.indent]is meaningless
```
Comment (tkrop) [inline chapters/http-status-codes-and-errors.adoc:402]: This looks like an unintended text fragment?

```suggestion
[.indent]
```

File: chapters/http-status-codes-and-errors.adoc [hunk 9]
```diff
@@ -366,53 +345,102 @@ The server understands the content type, but is unable to process the content.
...
+exponential back off pattern. If possible, the service should indicate how long
+the client should wait by setting the {Retry-After} header.
 
+[[status-code-504]]
+==== 504 Gateway Timeout {rfc-status-504} {use} {ALL}
+[.indent]
+The server, while acting as a gateway or proxy, did not receive a timely
+response. May be used by servers to indicate that an inbound service cannot
+process the request fast enough. Client may retry the request immediately
+exactly once, to check whether warming up the service solved the problem.
```
Comment (tkrop) [inline chapters/http-status-codes-and-errors.adoc:423]: `A client` or `Clients` I think.

File: chapters/http-status-codes-and-errors.adoc [hunk 10]
```diff
@@ -366,53 +345,102 @@ The server understands the content type, but is unable to process the content.
...
 
-[[status-code-503]]
-==== 503 Service Unavailable {rfc-status-503} {do-not-document} {ALL}
+[.do-not-use]
+[[status-code-505]]
+==== 505 HTTP Version Not Supported {rfc-status-505} {do-not-use} {ALL}
+[.indent]
+The server does not support the HTTP protocol version used in the request.
+Technical response code that serves not use case in RESTful APIs.
```
Comment (tkrop) [inline chapters/http-status-codes-and-errors.adoc:430]: Copilot is needs some polishing:

```suggestion
A technical response code that serves no use case in RESTful APIs.
```

**Review by tkrop — COMMENTED** [review]
@SmilyOrg have left some comments for polishing.

**Comment by SmilyOrg** [comment]
@tkrop this is ready for another look. Most of your comments were actually on a diff that was from a non-ideal merge, so I opened #795 for those. The diff should be cleaner now.

**Comment by tkrop** [comment]
👍

File: chapters/http-requests.adoc [hunk 11]
```diff
@@ -122,38 +122,47 @@ details and options.
...
+* For multiple resources you {SHOULD} return {201} in the response, as long as
+  they are created atomically, meaning that either all the resources are created
+  or none of them are.
+** You <<152>> if the request can partially fail, that is some of the resources
+  can fail to be created.
+* If the response does not contain the created resource or a resource monitor,
+  the status code {SHOULD} be {201} and the URL to {GET} the resource {MUST} be
+  provided in the {Location} response header.
+* If the {POST} is <<idempotent>>, the status code {SHOULD} be {201} if the
+  resource was created and {200} or {204} if the resource was updated.
```
Comment (ePaul) [inline chapters/http-requests.adoc:151]: I wonder whether these last two bullet points are more applicable for the single-resource case? They certainly don't fit for the "return 207" case. Maybe move them as sub-bullet-points to the single-resource case?

Comment (SmilyOrg) [reply chapters/http-requests.adoc:151]: Good point! I'll do that

**Comment by SmilyOrg** [comment]
@ePaul I added your suggestion, please check again!

**Comment by SmilyOrg** [comment]
:+1:

**Comment by ePaul** [comment]
:+1:
